Add fallback implementation for BLAS [sd]gemv_.

notoraptor opened the pull request:
Try to fix #6365 by adding a fallback implementation for `[sd]gemv_`, inspired from here: http://www.netlib.org/lapack/explore-html/d6/d30/group__single__blas__level2_gafc92361b74c6d41c7e5afa0aa5d13ec9.html#gafc92361b74c6d41c7e5afa0aa5d13ec9 . Tested successfully with `theano/tensor/nnet/tests/test_corr[3d]` with float32 and float64.

@abergeron @nouiz @lamblin 

PS: This PR would need to be fully tested on Windows.

notoraptor commented:
jenkins test windows

abergeron commented on theano/tensor/blas_headers.py:
```diff
@@ -834,7 +833,7 @@ def blas_header_text():
     /* Single Precision */
 
-        void sgemv_(char*, const int*, const int*, const float *, const float *, const int*, const float *, const int*, const float *, float *, const int*);
+        void sgemv_(char*, const int*, const int*, const float *, %(const)s float *, const int*, %(const)s float *, const int*, const float *, float *, const int*);
```
Why this change?

notoraptor replied:
Because I have to wrap A and x to PyArrayObjects, but NumPy functions don't take constant data pointers.

nouiz replied:
ok, it is the same as for gemm.

notoraptor commented:
The GPU card which runs jenkins on Windows does not support float16 precision: http://earlgrey.iro.umontreal.ca:8080/job/Theano_pr_win/38/testReport/junit/theano.gpuarray.tests/test_dnn/test_conv_guess_once_with_dtypes/

I will add a condition to failing test.

notoraptor commented:
Updated, test should be fixed now.

jenkins test windows

nouiz commented on theano/tensor/c_code/alt_blas_template.c:
```diff
@@ -66,25 +68,27 @@ PyObject* alt_op_%(float_type)s(int to_transpose, %(float_type)s* M, int nrow, i
-void %(name)s(
+
+/* gemm_ template code */
+void %(precision)sgemm_(
```
I think you want to remove the s in the name if you add precission here.

notoraptor replied:
In fact, the `s` is part of `%(precision)s`, so it is required to format the string :) !

nouiz replied:
right!

nouiz commented on theano/tensor/c_code/alt_blas_template.c:
```diff
@@ -66,25 +68,27 @@ PyObject* alt_op_%(float_type)s(int to_transpose, %(float_type)s* M, int nrow, i
 ) {
     if(*M < 0 || *N < 0 || *K < 0 || *LDA < 0 || *LDB < 0 || *LDC < 0)
-        alt_fatal_error("The integer arguments passed to %(name)s must all be at least 0.");
+        alt_fatal_error("The integer arguments passed to %(precision)sgemm_ must all be at least 0.");
```
idem

nouiz commented on theano/tensor/c_code/alt_blas_template.c:
```diff
@@ -156,19 +162,98 @@ void %(name)s(
             PyObject* alpha_opA_dot_opB = PyArray_Transpose((PyArrayObject*)opB_trans_dot_opA_trans, NULL);
             if(0 != PyArray_CopyInto((PyArrayObject*)matrix_C, (PyArrayObject*)alpha_opA_dot_opB))
-                alt_fatal_error("NumPy %(name)s implementation: unable to copy ALPHA*op(A)*op(B) into C when BETA == 0.");
+                alt_fatal_error("NumPy %(precision)sgemm_ implementation: unable to copy ALPHA*op(A)*op(B) into C when BETA == 0.");
```
idem

nouiz commented on theano/tensor/c_code/alt_blas_template.c:
```diff
@@ -156,19 +162,98 @@ void %(name)s(
 }
+
+/* gemv */
+void %(precision)sgemv_(
```
idem

nouiz commented on theano/tensor/blas_headers.py:
```diff
@@ -984,7 +983,11 @@ def blas_header_text():
+
+
+if not config.blas.ldflags:
+    print('Using NumPy based implementation for BLAS functions.')
```
Using NumPy C-API based ... So make it more clear that we still use C code.

But I'm not sure we should add this print expecially if on windows we don't link with anaconda mkl packages.

@lamblin what do you think?

nouiz commented on theano/tensor/c_code/alt_blas_template.c:
```diff
@@ -156,19 +162,98 @@ void %(name)s(
+    if (*M < 0 || *N < 0 || *LDA < 0 | *incx < 0 || *incy < 0)
+        alt_fatal_error("The integer arguments passed to %(precision)sgemv_ must all be at least 0.");
+    int transpose = alt_trans_to_bool(TRANS);
+    if (*M == 0 || *N == 0) {
```
I don't think this is always True. y size of M. So if M is 0, this is True. But if M is not like and N is zero, y isn,t 0. Remove `|| *N == 0` and update the code bellow and it is good I think.

Did you ran all tests under tensor/tests/test_blas*.py? in a config that don't have blas? Maybe you need to update some tests to don't be skipped.

notoraptor replied:
If we don't transpose A, then:
- `A * x` is of size `(M,N) * (N,1) => (M,1)`, and then y is expected to have size M.

But if we transpose A (ie. trans != 'n' and trans != 'N'), then:
- `A * x` is of size `(N,M) * (M,1) => (N,1)` and then y is expected to have size N.

So y should have size M (and `x` with size N) if we don't transpose, but y should have size N (and `x` with size M) if we transpose.

According to this documentation: http://www.netlib.org/lapack/explore-html/d6/d30/group__single__blas__level2_gafc92361b74c6d41c7e5afa0aa5d13ec9.html#gafc92361b74c6d41c7e5afa0aa5d13ec9

Indeed, there are skipped tests in `test_blas*`. I will activate them.

notoraptor replied:
All skipped tests check if BLAS C ops are inserted (eg. CGemv), but these ops seems to not be inserted when ldflags is empty, so tests will fail if they are not skipped.

Do we want to insert CGemv even if ldflags is empty (as it has now a fallback implementation) ?

nouiz replied:
yes. @lamblin do you agree with that?

nouiz replied:
In fact, I just got an hesitation. Can you at least do it locally to make sure all the tests pass?
I'm not convinced the c code will be faster then the python code that call blas. The python code have more overhead, but a direct call to blas could be faster

nouiz replied:
We are so close to a release. I prefer to don't enable CGemv in those case by default for now. But can you enable it locally and run the tests to make sure the C code is good?

notoraptor replied:
Ok, I will test it locally! Report to coming.

notoraptor replied:
With ldflags empty, CGemv doesn't compile because its needs `[sd]dot_`, which does not have fallback C code yet !

nouiz replied:
Can you quickly modify the cgemv code to raise an error in that case so that it compile or quickly implement it in the op? I don't like having new c code with code path not tested. The conv code don't tests all the code path you added.

notoraptor replied:
Anyway, this fallback code will be tested with `theano.tensor.nnet.corr` Ops, as they now use [sd]gemv_ (because of unshared convolutions) and they have to compile whatever ldflags contains.

With ldflags empty, this test passes on my computer (Windows 10 64 bits) and should pass with jenkins on windows too:
```
nosetests -xvs theano\tensor\nnet\tests\test_corr.py
```
```
(python2) C:\Users\notoraptor\mila\dev\git\theano>nosetests -xvs theano\tensor\nnet\tests\test_corr.py
Using NumPy C-API based implementation for BLAS functions.
test_fwd (theano.tensor.nnet.tests.test_corr.Grouped_conv_noOptim) ... ok
test_gradinputs (theano.tensor.nnet.tests.test_corr.Grouped_conv_noOptim) ... ok
test_gradweights (theano.tensor.nnet.tests.test_corr.Grouped_conv_noOptim) ... ok
Tests that basic correlations work for odd and even ... ok
Checks dtype upcast for CorrMM methods. ... ok
Tests correlation where filter dilation != (1,1) ... ok
Tests basic correlation in full mode and case where filter ... ok
test_img_kernel_same_shape (theano.tensor.nnet.tests.test_corr.TestCorr2D) ... ok
test_infer_shape_forward (theano.tensor.nnet.tests.test_corr.TestCorr2D) ... ok
test_infer_shape_gradI (theano.tensor.nnet.tests.test_corr.TestCorr2D) ... ok
test_infer_shape_gradW (theano.tensor.nnet.tests.test_corr.TestCorr2D) ... ok
Tests scenario where filter_shape[1] != input_shape[1] ... ok
test_non_contiguous (theano.tensor.nnet.tests.test_corr.TestCorr2D) ... ok
Tests correlation where the {image,filter}_shape is a Constant tensor. ... ok
Tests correlation where subsampling != (1,1) ... ok
Make sure errors are raised when image and kernel are not 4D tensors ... ok
test_fwd (theano.tensor.nnet.tests.test_corr.TestGroupCorr2d) ... ok
test_gradinputs (theano.tensor.nnet.tests.test_corr.TestGroupCorr2d) ... ok
test_gradweights (theano.tensor.nnet.tests.test_corr.TestGroupCorr2d) ... ok
test_graph (theano.tensor.nnet.tests.test_corr.TestGroupCorr2d) ... ok
test_fwd (theano.tensor.nnet.tests.test_corr.TestUnsharedConv) ... ok
test_gradinput (theano.tensor.nnet.tests.test_corr.TestUnsharedConv) ... ok
test_gradweight (theano.tensor.nnet.tests.test_corr.TestUnsharedConv) ... ok
test_fwd (theano.tensor.nnet.tests.test_corr.TestUnsharedCorr2d) ... ok
test_gradinput (theano.tensor.nnet.tests.test_corr.TestUnsharedCorr2d) ... ok
test_gradweight (theano.tensor.nnet.tests.test_corr.TestUnsharedCorr2d) ... ok

----------------------------------------------------------------------
Ran 26 tests in 90.612s

OK
```

nouiz replied:
This tests some of the code path of the new code, not all of them. For example, I'm pretty alpha/beta is always zeros or non-zeros. It will probably only test with contiguous inputs too.

notoraptor replied:
Ok, I see.

> Can you quickly modify the cgemv code to raise an error in that case so that it compile or quickly implement it in the op?

I don't hink I can modify cgemv code to raise an error if [sd]dot_ is not available (or maybe by implementing a fallback [sd]dot_ that always generates an error when it is called ?).

I think I should better make another PR to fix MKL linking, and then add a full implementation of [sd]dot_ to this current PR later (maybe tomorrow ?), after we made the beta2. So, we won't merge this current PR until it can be fully tested.

notoraptor commented:
I think I have ultimately fixed MKL linking ! It was just a typo to fix: https://github.com/Theano/Theano/pull/6367/commits/f086ca178345bd89f61c26293d7a572ef9b04d39 . GCC is normally able to link with DLLs ( https://cygwin.com/cygwin-ug-net/dll.html ).

When testing `python -m theano.misc.check_blas > log-python2.log 2>&1`:
```

        Some results that you can compare against. They were 10 executions
        of gemm in float64 with matrices of shape 2000x2000 (M=N=K=2000).
        All memory layout was in C order.

        CPU tested: Xeon E5345(2.33Ghz, 8M L2 cache, 1333Mhz FSB),
                    Xeon E5430(2.66Ghz, 12M L2 cache, 1333Mhz FSB),
                    Xeon E5450(3Ghz, 12M L2 cache, 1333Mhz FSB),
                    Xeon X5560(2.8Ghz, 12M L2 cache, hyper-threads?)
                    Core 2 E8500, Core i7 930(2.8Ghz, hyper-threads enabled),
                    Core i7 950(3.07GHz, hyper-threads enabled)
                    Xeon X5550(2.67GHz, 8M l2 cache?, hyper-threads enabled)


        Libraries tested:
            * numpy with ATLAS from distribution (FC9) package (1 thread)
            * manually compiled numpy and ATLAS with 2 threads
            * goto 1.26 with 1, 2, 4 and 8 threads
            * goto2 1.13 compiled with multiple threads enabled

                          Xeon   Xeon   Xeon  Core2 i7    i7     Xeon   Xeon
        lib/nb threads    E5345  E5430  E5450 E8500 930   950    X5560  X5550

        numpy 1.3.0 blas                                                775.92s
        numpy_FC9_atlas/1 39.2s  35.0s  30.7s 29.6s 21.5s 19.60s
        goto/1            18.7s  16.1s  14.2s 13.7s 16.1s 14.67s
        numpy_MAN_atlas/2 12.0s  11.6s  10.2s  9.2s  9.0s
        goto/2             9.5s   8.1s   7.1s  7.3s  8.1s  7.4s
        goto/4             4.9s   4.4s   3.7s  -     4.1s  3.8s
        goto/8             2.7s   2.4s   2.0s  -     4.1s  3.8s
        openblas/1                                        14.04s
        openblas/2                                         7.16s
        openblas/4                                         3.71s
        openblas/8                                         3.70s
        mkl 11.0.083/1            7.97s
        mkl 10.2.2.025/1                                         13.7s
        mkl 10.2.2.025/2                                          7.6s
        mkl 10.2.2.025/4                                          4.0s
        mkl 10.2.2.025/8                                          2.0s
        goto2 1.13/1                                                     14.37s
        goto2 1.13/2                                                      7.26s
        goto2 1.13/4                                                      3.70s
        goto2 1.13/8                                                      1.94s
        goto2 1.13/16                                                     3.16s

        Test time in float32. There were 10 executions of gemm in
        float32 with matrices of shape 5000x5000 (M=N=K=5000)
        All memory layout was in C order.


        cuda version      8.0    7.5    7.0
        gpu
        M40               0.45s  0.47s
        k80               0.92s  0.96s
        K6000/NOECC       0.71s         0.69s
        P6000/NOECC       0.25s

        Titan X (Pascal)  0.28s
        GTX Titan X       0.45s  0.45s  0.47s
        GTX Titan Black   0.66s  0.64s  0.64s
        GTX 1080          0.35s
        GTX 980 Ti               0.41s
        GTX 970                  0.66s
        GTX 680                         1.57s
        GTX 750 Ti               2.01s  2.01s
        GTX 750                  2.46s  2.37s
        GTX 660                  2.32s  2.32s
        GTX 580                  2.42s
        GTX 480                  2.87s
        TX1                             7.6s (float32 storage and computation)
        GT 610                          33.5s
        
Some Theano flags:
    blas.ldflags= -L"C:\Users\notoraptor\data\programs\Anaconda3\envs\python2\Library\bin" -lmkl_core -lmkl_intel_thread -lmkl_rt
    compiledir= C:\Users\notoraptor\AppData\Local\Theano\compiledir_Windows-10-10.0.15063-Intel64_Family_6_Model_158_Stepping_9_GenuineIntel-2.7.13-64
    floatX= float64
    device= cpu
Some OS information:
    sys.platform= win32
    sys.version= 2.7.13 |Continuum Analytics, Inc.| (default, May 11 2017, 13:17:26) [MSC v.1500 64 bit (AMD64)]
    sys.prefix= C:\Users\notoraptor\data\programs\Anaconda3\envs\python2
Some environment variables:
    MKL_NUM_THREADS= None
    OMP_NUM_THREADS= None
    GOTO_NUM_THREADS= None

Numpy config: (used when the Theano flag "blas.ldflags" is empty)
lapack_opt_info:
    libraries = ['mkl_core_dll', 'mkl_intel_lp64_dll', 'mkl_intel_thread_dll']
    library_dirs = ['C:/Users/notoraptor/data/programs/Anaconda3/envs/python2\\Library\\lib']
    define_macros = [('SCIPY_MKL_H', None), ('HAVE_CBLAS', None)]
    include_dirs = ['C:/Users/notoraptor/data/programs/Anaconda3/envs/python2\\Library\\include']
blas_opt_info:
    libraries = ['mkl_core_dll', 'mkl_intel_lp64_dll', 'mkl_intel_thread_dll']
    library_dirs = ['C:/Users/notoraptor/data/programs/Anaconda3/envs/python2\\Library\\lib']
    define_macros = [('SCIPY_MKL_H', None), ('HAVE_CBLAS', None)]
    include_dirs = ['C:/Users/notoraptor/data/programs/Anaconda3/envs/python2\\Library\\include']
lapack_mkl_info:
    libraries = ['mkl_core_dll', 'mkl_intel_lp64_dll', 'mkl_intel_thread_dll']
    library_dirs = ['C:/Users/notoraptor/data/programs/Anaconda3/envs/python2\\Library\\lib']
    define_macros = [('SCIPY_MKL_H', None), ('HAVE_CBLAS', None)]
    include_dirs = ['C:/Users/notoraptor/data/programs/Anaconda3/envs/python2\\Library\\include']
blas_mkl_info:
    libraries = ['mkl_core_dll', 'mkl_intel_lp64_dll', 'mkl_intel_thread_dll']
    library_dirs = ['C:/Users/notoraptor/data/programs/Anaconda3/envs/python2\\Library\\lib']
    define_macros = [('SCIPY_MKL_H', None), ('HAVE_CBLAS', None)]
    include_dirs = ['C:/Users/notoraptor/data/programs/Anaconda3/envs/python2\\Library\\include']
Numpy dot module: numpy.core.multiarray
Numpy location: C:\Users\notoraptor\data\programs\Anaconda3\envs\python2\lib\site-packages\numpy\__init__.pyc
Numpy version: 1.13.1

We executed 10 calls to gemm with a and b matrices of shapes (5000, 5000) and (5000, 5000).

Total execution time: 17.67s on CPU (with direct Theano binding to blas).

Try to run this script a few times. Experience shows that the first time is not as fast as followings calls. The difference is not big, but consistent.

```

Now ldflags is not empty !
```
Some Theano flags:
    blas.ldflags= -L"C:\Users\notoraptor\data\programs\Anaconda3\envs\python2\Library\bin" -lmkl_core -lmkl_intel_thread -lmkl_rt
```

About skipped tests, they always check if CGemv is in the graph, but CGemv is currently inserted only if ldflags is not empty. Do we want to always insert CGemv (as there is now a fallback implementation) ?

@abergeron @nouiz

notoraptor commented:
On Windows, Jenkins had skipped some tests because ldflags was empty: http://earlgrey.iro.umontreal.ca:8080/job/Theano_pr_win/38/testReport/theano.tensor.tests.test_blas_c/ . Now these tests should be run.

jenkins test windows

notoraptor commented:
Updated.
- I added fallback implementation for `[sd]dot_`.
- I removed parametrization of `const` keyword in blas header, as it seems C/C++ ignores `const` keyword in functions declarations (it's allowed to declare a parameter as `const` in function declarion but non const in function definition: https://stackoverflow.com/questions/3681188/why-does-a-function-declaration-with-a-const-argument-allow-calling-of-a-functio ). And it compiles as-is!
```
(python2) C:\Users\notoraptor\mila\dev\git\theano>nosetests -xvs theano\tensor\tests\test_blas_c.py:TestCGemv
Using NumPy C-API based implementation for BLAS functions.
test_force_gemv_init (theano.tensor.tests.test_blas_c.TestCGemv) ... ok
test_gemv1 (theano.tensor.tests.test_blas_c.TestCGemv) ... ok
test_gemv_dimensions (theano.tensor.tests.test_blas_c.TestCGemv) ... ok
test_multiple_inplace (theano.tensor.tests.test_blas_c.TestCGemv) ... ok
test_nan_beta_0 (theano.tensor.tests.test_blas_c.TestCGemv) ... ok
Test matrix dot vector ... ok
Test vector dot matrix ... ok

----------------------------------------------------------------------
Ran 7 tests in 2.126s

OK
```

@nouiz @abergeron

nouiz commented on theano/tensor/c_code/alt_blas_template.c:
```diff
@@ -257,3 +259,33 @@ void %(precision)sgemv_(
+        alt_fatal_error("NumPy %(precision)sdot_: unables to compute dot.");
+    // Get result.
+    %(float_type)s result = *(%(float_type)s*)PyArray_DATA((PyArrayObject*)dot_product);
+    Py_XDECREF(dot_product);
```
This will free the prt result you keep. I think you need to use MatrixProduct2() but allocate the output with malloc first, then wrap it in PyArray and pass it to MatrixProduct2

notoraptor replied:
I don't keep the pointer, I dereference and copy value instead (`%(float_type)s result = *(` `...`), as we expect just 1 scalar value.

lamblin commented on theano/tensor/blas_headers.py:
```diff
@@ -984,7 +981,11 @@ def blas_header_text():
+
+
+if not config.blas.ldflags:
+    print('Using NumPy C-API based implementation for BLAS functions.')
```
Could it be possible to use the `logging` module here instead?

notoraptor replied:
Done!

notoraptor commented:
Updated. Added tests to explicitely check fallback for `gemv` and `dot` in any case, and update code to fix them (support negative strides) and clear code.

@nouiz @lamblin @abergeron

notoraptor commented:
All tests pass here !

notoraptor commented:
Rebased to integrate changes merged from #6371 .
